### PR TITLE
Partition data tx calc predicate

### PR DESCRIPTION
**Pull request by FloatingCrowbar**
LOGBROKER-8900

### CODE REVIEW

File: ydb/core/persqueue/partition.cpp [hunk 1]
```diff
@@ -1009,19 +1009,54 @@ void TPartition::Handle(TEvPQ::TEvGetWriteInfoRequest::TPtr& ev, const TActorCon
...
+void TPartition::Handle(TEvPQ::TEvGetWriteInfoResponse::TPtr& ev, const TActorContext&) {
+    auto& srcIdInfo = ev->Get()->SrcIdInfo;
+
+    Y_ABORT_UNLESS(TxInProgress);
+    auto& knownSrcIds = SourceIdStorage.GetInMemorySourceIds();
+    auto& currTx = GetUserActionAndTransactionEventsFront<TTransaction>();
+
+    for (auto& s : srcIdInfo) {
+        auto ins = currTx.AffectedSourcesIds.insert(s.first).second;
+        if (!ins) {
```
Comment (Alek5andr-Kotov) [inline ydb/core/persqueue/partition.cpp:1031]: В каком случае `ans == false`? AffectedSourcesIds это `THashSet<>`, а `srcIdInfo` это `THashMap<>`

File: ydb/core/persqueue/partition.cpp [hunk 2]
```diff
@@ -1009,19 +1009,54 @@ void TPartition::Handle(TEvPQ::TEvGetWriteInfoRequest::TPtr& ev, const TActorCon
...
+}
+
+void TPartition::Handle(TEvPQ::TEvGetWriteInfoError::TPtr&, const TActorContext&) {
+    Y_ABORT_UNLESS(TxInProgress);
+    auto& currTx = GetUserActionAndTransactionEventsFront<TTransaction>();
+    return SendCalcPredicateResult(currTx, false);
+}
+
+void TPartition::SendCalcPredicateResult(TTransaction& tx, bool result) {
+    tx.Predicate = result;
```
Comment (Alek5andr-Kotov) [inline ydb/core/persqueue/partition.cpp:1053]: Может стоит переименовать функцию? Я не ожидал, что она не только отправляет ответ, но и меняет состояние транзакции

File: ydb/core/persqueue/partition.cpp [hunk 3]
```diff
@@ -1640,13 +1675,15 @@ bool TPartition::ProcessUserActionOrTransaction(TTransaction& t,
...
-        ctx.Send(Tablet,
-                 MakeHolder<TEvPQ::TEvTxCalcPredicateResult>(t.Tx->Step,
-                                                             t.Tx->TxId,
-                                                             Partition,
-                                                             *t.Predicate).Release());
-
+        if (t.Tx->SupportivePartitionActor) {
+            Send(t.Tx->SupportivePartitionActor, new TEvPQ::TEvGetWriteInfoRequest());
+        } else {
+            ctx.Send(Tablet,
```
Comment (Alek5andr-Kotov) [inline ydb/core/persqueue/partition.cpp:1681]: Давай здесь тоже использовать метод `Send`?

File: ydb/core/persqueue/ut/partition_ut.cpp [hunk 4]
```diff
@@ -363,7 +364,14 @@ TPartition* TPartitionFixture::CreatePartition(const TCreatePartitionParams& par
...
         SendConfigResponse(params.Config);
     } else {
-        ret = CreatePartitionActor(params.Partition, config, false, params.Transactions);
+        TVector<TTransaction> copyTx;
+        for (const auto& origTx : params.Transactions) {
+            copyTx.emplace_back(origTx.Tx, origTx.Predicate);
+            copyTx.back().ChangeConfig = origTx.ChangeConfig;
+            copyTx.back().SendReply = origTx.SendReply;
+            copyTx.back().ProposeConfig = origTx.ProposeConfig;
+        }
```
Comment (Alek5andr-Kotov) [inline ydb/core/persqueue/ut/partition_ut.cpp:373]: Зачем нужно вручную создавать копию?

File: ydb/core/persqueue/ut/partition_ut.cpp [hunk 5]
```diff
@@ -598,9 +605,30 @@ void TPartitionFixture::SendChangeOwner(const ui64 cookie, const TString& owner,
 }
 
 void TPartitionFixture::SendGetWriteInfo(const ui32 internalPartitionId) {
-    auto event = MakeHolder<TEvPQ::TEvGetWriteInfoRequest>(internalPartitionId);
+    Y_UNUSED(internalPartitionId);
+    auto event = MakeHolder<TEvPQ::TEvGetWriteInfoRequest>();
```
Comment (Alek5andr-Kotov) [inline ydb/core/persqueue/ut/partition_ut.cpp:609]: Предлагаю совсем удалить параметр `internalPartitionId`

File: ydb/core/persqueue/partition.cpp [hunk 6]
```diff
@@ -1640,13 +1675,15 @@ bool TPartition::ProcessUserActionOrTransaction(TTransaction& t,
...
     if (t.Tx) {
         t.Predicate = BeginTransaction(*t.Tx, ctx);
-
-        ctx.Send(Tablet,
-                 MakeHolder<TEvPQ::TEvTxCalcPredicateResult>(t.Tx->Step,
-                                                             t.Tx->TxId,
-                                                             Partition,
-                                                             *t.Predicate).Release());
-
+        if (t.Tx->SupportivePartitionActor) {
```
Comment (Alek5andr-Kotov) [inline ydb/core/persqueue/partition.cpp:1678]: Могут быть транзакции в которых есть и чтение, и запись. Сейчас не учитывается значение предиката, которое вернула функция `BeginTransaction`.

File: ydb/core/persqueue/partition.cpp [hunk 7]
```diff
@@ -1009,19 +1009,54 @@ void TPartition::Handle(TEvPQ::TEvGetWriteInfoRequest::TPtr& ev, const TActorCon
...
 
     ctx.Send(ev->Sender, response);
 }
 
+void TPartition::Handle(TEvPQ::TEvGetWriteInfoResponse::TPtr& ev, const TActorContext&) {
+    auto& srcIdInfo = ev->Get()->SrcIdInfo;
+
+    Y_ABORT_UNLESS(TxInProgress);
+    auto& knownSrcIds = SourceIdStorage.GetInMemorySourceIds();
+    auto& currTx = GetUserActionAndTransactionEventsFront<TTransaction>();
```
Comment (Alek5andr-Kotov) [inline ydb/core/persqueue/partition.cpp:1027]: Лучше `GetCurrentTransaction`
https://github.com/ydb-platform/ydb/blob/main/ydb/core/persqueue/partition.h#L669

File: ydb/core/persqueue/ut/partition_ut.cpp (hunk 4 above)
Comment (FloatingCrowbar) [reply ydb/core/persqueue/ut/partition_ut.cpp:373]: Обсудили лично. Из-за того, что появился Holder, автоматическое копирование не работает.

File: ydb/core/persqueue/partition.h [hunk 8]
```diff
@@ -65,8 +66,11 @@ struct TTransaction {
 
     TSimpleSharedPtr<TEvPQ::TEvChangePartitionConfig> ChangeConfig;
     bool SendReply;
-
     TSimpleSharedPtr<TEvPQ::TEvProposePartitionConfig> ProposeConfig;
+
+    //Data Tx
+    THashSet<TString> AffectedSourcesIds;
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.h:72]: Why do we need this? I thought we will use same object for storing changed source Ids and their seq no with usual write.

File: ydb/core/persqueue/partition.cpp [hunk 9]
```diff
@@ -1640,24 +1674,26 @@ bool TPartition::ProcessUserActionOrTransaction(TTransaction& t,
...
     if (t.Tx) {
         t.Predicate = BeginTransaction(*t.Tx, ctx);
-
-        ctx.Send(Tablet,
-                 MakeHolder<TEvPQ::TEvTxCalcPredicateResult>(t.Tx->Step,
-                                                             t.Tx->TxId,
-                                                             Partition,
-                                                             *t.Predicate).Release());
-
+        if (t.Tx->SupportivePartitionActor && t.Predicate.GetOrElse(true)) {
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:1677]: BeginTransaction will set t.Predicate to true in case of Tx->SupportivePartitionActor. Always. Why do we check t.Predicate here?

File: ydb/core/persqueue/partition.cpp [hunk 10]
```diff
@@ -1624,113 +1618,168 @@ size_t TPartition::GetUserActCount(const TString& consumer) const
 
 void TPartition::ProcessTxsAndUserActs(const TActorContext& ctx)
 {
-    if (KVWriteInProgress || TxInProgress) {
+    if (KVWriteInProgress) {
         return;
     }
+    bool shouldGoOn = true;
+    while (shouldGoOn) {
+        if (WaitingForCommit) {
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:1626]: WaitingForTxCommitOrRollback

File: ydb/core/persqueue/partition.cpp [hunk 11]
```diff
@@ -1624,113 +1618,168 @@ size_t TPartition::GetUserActCount(const TString& consumer) const
...
+        return this->ExecUserActionOrTransaction(event, request, commit);
+    };
+    while (!UserActionAndTxPendingCommit.empty()) {
+        auto& front = UserActionAndTxPendingCommit.front();
+        auto& event = front.Event;
+        switch (front.State) {
+            case ECommitState::Pending:
+                return;
+            case ECommitState::Aborted:
+                commit = false;
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:1715]: Why is commit global variable, not per UserActionAndTxPendingCommit?

File: ydb/core/persqueue/partition.cpp [hunk 12]
```diff
@@ -1624,113 +1618,168 @@ size_t TPartition::GetUserActCount(const TString& consumer) const
...
+                break;
+            case ECommitState::Committed:
+                break;
+        }
+        std::visit(visitor, event);
+
+        UserActionAndTxPendingCommit.pop_front();
+    }
+    WriteInfosToTx.clear();
+    TxAffectedClients.clear();
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:1725]: TxAffectedConsumers

File: ydb/core/persqueue/partition.cpp [hunk 13]
```diff
@@ -1624,113 +1618,168 @@ size_t TPartition::GetUserActCount(const TString& consumer) const
...
+                commit = false;
+                break;
+            case ECommitState::Committed:
+                break;
+        }
+        std::visit(visitor, event);
+
+        UserActionAndTxPendingCommit.pop_front();
+    }
+    WriteInfosToTx.clear();
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:1724]: Would not you clear new transactions that not in this batch?

File: ydb/core/persqueue/partition.cpp [hunk 14]
```diff
@@ -1081,19 +988,91 @@ void TPartition::Handle(TEvPQ::TEvGetWriteInfoRequest::TPtr& ev, const TActorCon
...
+
+    EProcessResult ret = EProcessResult::ContinueWaitCommit;
+    const auto& knownSourceIds = SourceIdStorage.GetInMemorySourceIds();
+    for (auto& s : srcIdInfo) {
+        auto ins = TxAffectedSourcesIds.insert(s.first).second;
+        if (!ins) {
+            ret = EProcessResult::Blocked;
+            break;
+        }
+        if (WriteAffectedSourcesIds.contains(s.first)) {
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:1029]: WriteAffectedSourceIds never is set. Is it right?

File: ydb/core/persqueue/partition.cpp [hunk 15]
```diff
@@ -2258,83 +2296,101 @@ void TPartition::ProcessImmediateTx(const NKikimrPQ::TEvProposeTransaction& tx,
...
+) {
     Y_ABORT_UNLESS(!KVWriteInProgress);
 
     const TString& user = act.ClientId;
-    const bool strictCommitOffset = (act.Type == TEvPQ::TEvSetClientInfo::ESCI_OFFSET && act.Strict);
+    if (act.Type == TEvPQ::TEvSetClientInfo::ESCI_OFFSET) {
+        if (TxAffectedClients.contains(user)) {
+            return EProcessResult::Blocked;
+        }
+        SetOffsetAffectedClients.insert(user);
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:2384]: why do SetOffsetAffectedClient.insert called twice - here and 2 lines after?

File: ydb/core/persqueue/partition.cpp [hunk 16]
```diff
@@ -2258,83 +2296,101 @@ void TPartition::ProcessImmediateTx(const NKikimrPQ::TEvProposeTransaction& tx,
...
+    }
+    SetOffsetAffectedClients.insert(user);
+    return EProcessResult::ContinueAutoCommit;
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:2387]: What do ContiueAutoCommit means? And when must be used?

File: ydb/core/persqueue/partition.cpp [hunk 17]
```diff
@@ -1081,19 +988,91 @@ void TPartition::Handle(TEvPQ::TEvGetWriteInfoRequest::TPtr& ev, const TActorCon
...
+}
+
+TPartition::EProcessResult TPartition::ApplyWriteInfoResponse(TTransaction& tx) {
+    Y_ABORT_UNLESS(tx.WriteInfo);
+    auto& srcIdInfo = tx.WriteInfo->SrcIdInfo;
+
+    EProcessResult ret = EProcessResult::ContinueWaitCommit;
+    const auto& knownSourceIds = SourceIdStorage.GetInMemorySourceIds();
+    for (auto& s : srcIdInfo) {
+        auto ins = TxAffectedSourcesIds.insert(s.first).second;
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:1024]: In case of immediateTx you need to use WriteAffectedSourceIds. The same bug with TxAffectedClients.

**Review by alexnick88 — DISMISSED** [review]

File: ydb/core/persqueue/partition.cpp (hunk 11 above)
Comment (FloatingCrowbar) [reply ydb/core/persqueue/partition.cpp:1715]: This is the state of single UserAction.

File: ydb/core/persqueue/partition.cpp (hunk 16 above)
Comment (FloatingCrowbar) [reply ydb/core/persqueue/partition.cpp:2387]: No additional TxCommit signal will be expected for this action, it can be processed instantly as soon as we get it from the queue. Used for anything except transactions.

File: ydb/core/persqueue/partition.cpp [hunk 18]
```diff
@@ -654,8 +625,6 @@ void TPartition::Handle(TEvPQ::TEvPipeDisconnected::TPtr& ev, const TActorContex
 void TPartition::Handle(TEvPQ::TEvPartitionStatus::TPtr& ev, const TActorContext& ctx) {
     NKikimrPQ::TStatusResponse::TPartResult result;
     result.SetPartition(Partition.InternalPartitionId);
-    result.SetGeneration(TabletGeneration);
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:657]: Why did you remove it?

File: ydb/core/persqueue/partition.cpp [hunk 19]
```diff
@@ -979,114 +944,46 @@ void TPartition::Handle(TEvPQ::TEvTxCalcPredicate::TPtr& ev, const TActorContext
...
     }
 
-    auto& t = GetCurrentTransaction();
-    Y_ABORT_UNLESS(t.Tx);
-
-    if (WriteInfoResponse->BodyKeys.empty()) {
-        t.Predicate = predicate && BeginTransaction(*t.Tx, ctx);
+    auto txIter = TransactionsInflight.begin();
+    if (ChangeConfig) {
+        Y_ABORT_UNLESS(TransactionsInflight.size() == 1);
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:956]: How do you guarantee it?

File: ydb/core/persqueue/ut/partition_ut.cpp [hunk 20]
```diff
@@ -1970,10 +2462,606 @@ Y_UNIT_TEST_F(ShadowPartitionCountersRestore, TPartitionFixture) {
...
+    SendWriteInfoResponse(tx3);
+
+    ExpectNoBatchCompletion();
+    SendWriteInfoResponse(tx1);
+    WaitTxPredicateReply(tx1);
+    SendWriteInfoResponse(tx2);
+    WaitTxPredicateReply(tx2);
+    WaitTxPredicateReply(tx3);
+
+    WaitBatchCompletion(5 + 6 + 6); //5 tx or immediate tx + 2 normal writes with 2 messages each;
```
[the remark this answers is not in the document]
Comment (FloatingCrowbar) [reply ydb/core/persqueue/ut/partition_ut.cpp:2577]: Each normal write produces separate UserAction for each message. NormalWrite with SeqNo range = 7-12 writes 6 messages and generates 6 actions.

File: ydb/core/persqueue/ut/partition_ut.cpp [hunk 21]
```diff
@@ -1970,10 +2462,606 @@ Y_UNIT_TEST_F(ShadowPartitionCountersRestore, TPartitionFixture) {
...
+    TTxBatchingTestParams params {.WriterSessions{"src1", "src4"},.EndOffset=1};
+    Init(std::move(params));
+
+    auto tx1 = MakeAndSendWriteTx({{"src1", {1, 3}}});
+    auto tx2 = MakeAndSendWriteTx({{"src2", {4, 6}}});
+    auto tx3 = MakeAndSendWriteTx({{"src1", {4, 6}}});
+
+    AddAndSendNormalWrite("src1", 7, 12);
+    AddAndSendNormalWrite("src4", 1, 2);
+    auto tx5 = MakeAndSendWriteTx({{"src4", {4, 5}}});
```
[the remark this answers is not in the document]
Comment (FloatingCrowbar) [reply ydb/core/persqueue/ut/partition_ut.cpp:2601]: Just naming. There was tx4 before but was removed during test refactory. Can rename it

File: ydb/core/persqueue/partition.cpp [hunk 22]
```diff
@@ -1108,19 +1005,116 @@ void TPartition::Handle(TEvPQ::TEvGetWriteInfoRequest::TPtr& ev, const TActorCon
...
+            tx.Message = TStringBuilder() << "MinSeqNo violation failure on " << s.first;
+            tx.WriteInfoApplied = true;
+            break;
+        }
+    }
+    if (ret == EProcessResult::Continue && tx.Predicate.GetOrElse(true)) {
+        TxAffectedSourcesIds.insert(txSourceIds.begin(), txSourceIds.end());
+        tx.WriteInfoApplied = true;
+        WriteKeysSizeEstimate += tx.WriteInfo->BodyKeys.size();
+        WriteKeysSizeEstimate += tx.WriteInfo->SrcIdInfo.size();
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:1081]: Add here BlboFromHead length

File: ydb/core/persqueue/ut/partition_ut.cpp [hunk 23]
```diff
@@ -1970,10 +2462,606 @@ Y_UNIT_TEST_F(ShadowPartitionCountersRestore, TPartitionFixture) {
...
+    WaitBatchCompletion(6 + 2); // Normal writes produce 1 act for each message
+    ExpectNoTxPredicateReply();
+    WaitKvRequest();
+    SendKvResponse();
+    WaitCommitDone(tx1);
+    WaitCommitDone(tx3);
+    WaitTxPredicateReply(tx5);
+    WaitBatchCompletion(1);
+    SendTxCommit(tx5);
+    WaitBatchCompletion(1 + 6); //Normal write & immTx for src4;
```
[the remark this answers is not in the document]
Comment (FloatingCrowbar) [reply ydb/core/persqueue/ut/partition_ut.cpp:2633]: Same, 6 messages from normal write produce 6 user actions.

File: ydb/core/persqueue/partition.cpp [hunk 24]
```diff
@@ -1649,127 +1655,206 @@ size_t TPartition::GetUserActCount(const TString& consumer) const
...
+        }
+        BatchingState = ETxBatchingState::PreProcessing;
+    }
+}
 
+void TPartition::ContinueProcessTxsAndUserActs(const TActorContext&)
+{
+    Y_ABORT_UNLESS(!KVWriteInProgress);
+
+    if (WriteCycleSizeEstimate >= MAX_WRITE_CYCLE_SIZE || WriteKeysSizeEstimate >= MAX_USERS) {
```
Comment (alexnick88) [inline ydb/core/persqueue/partition.cpp:1715]: Rename MAX_USERS to MAX_KEYS. and change it to 10k. And check that this limit is bigger than limit for data blobs in supportive partition. There should be some limit in it.

File: ydb/core/persqueue/ut/partition_ut.cpp [hunk 25]
```diff
@@ -1970,10 +2462,606 @@ Y_UNIT_TEST_F(ShadowPartitionCountersRestore, TPartitionFixture) {
...
+    {
+    // 2. ImmTx -> WriteTx = KVRequest
+    ResetBatchCompletion();
+    wrapper.Start();
+    auto immTx = wrapper.AddImmediateTx();
+    auto tx = wrapper.AddTx();
+    wrapper.Process();
+    WaitWriteInfoRequest(immTx, true);
+    WaitWriteInfoRequest(tx, true);
+    WaitBatchCompletion(1+1);
```
[the remark this answers is not in the document]
Comment (FloatingCrowbar) [reply ydb/core/persqueue/ut/partition_ut.cpp:2769]: That's 1 for TxTmp created in wrapper to block execution + 1 for immTx. They are batched because TxTmp has no affected SrcIds.

File: ydb/core/persqueue/partition.cpp (hunk 19 above)
Comment (FloatingCrowbar) [reply ydb/core/persqueue/partition.cpp:956]: ChangeConfig is blocked if FirstEvent flag is non set. And anything else is blocked if ChangeConfig is defined.

**Review by alexnick88 — DISMISSED** [review]

**Review by alexnick88 — DISMISSED** [review]

**Review by alexnick88 — APPROVED** [review]
